### PR TITLE
#620: Make Lineage reactive

### LINKED ISSUE #620: Refactor Lineage to the reactive approach

### PR DESCRIPTION
**Pull request by esmeneev**
**What changes did you make?** (Give an overview)
Closes #620 
Refactored `LineageRepository`, `GroupEntityRelationRepository` and `LineageService` to reactive approach.

Tests and refactoring of `LineageMapper` still pending.

**Is there anything you'd like reviewers to focus on?**

**How to test**
Will be covered by unit tests, also can be tested by calling respective endpoints (`GET /api/dataentities/{data_entity_id}/lineage/downstream`, `GET /api/dataentities/{data_entity_id}/lineage/upstream` and `GET /api/dataentitygroups/{data_entity_group_id}/lineage`)

### CODE REVIEW

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/service/LineageServiceImpl.java [hunk 1]
```diff
@@ -41,91 +42,103 @@
...
+                .collectList()
+                .flatMap(relations -> {
+                    final Set<String> oddrnsToFetch = relations.stream()
+                        .flatMap(r -> Stream.of(r.getParentOddrn(), r.getChildOddrn()))
+                        .collect(Collectors.toSet());
+                    return Mono
+                        .zip(groupEntityRelationRepository.fetchGroupRelations(oddrnsToFetch),
+                            lineageRepository.getChildrenCount(oddrnsToFetch),
+                            lineageRepository.getParentCount(oddrnsToFetch))
+                        .map(tuple -> {
```
Comment (damirabdul) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/service/LineageServiceImpl.java:106]: you can use TupleUtils.finction() for giving names to tuples components. There are a lot of examples in codebase

Comment (damirabdul) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/service/LineageServiceImpl.java:106]: Can we extract peaces of code from this map to small methods?

**Review by damirabdul — CHANGES_REQUESTED** [review]
Please don't forget to delete unused methods from blocking repositories

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveLineageRepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+@Repository
+@RequiredArgsConstructor
+public class ReactiveLineageRepositoryImpl implements ReactiveLineageRepository {
+
+    private final JooqReactiveOperations jooqReactiveOperations;
+
+    @Override
+    @ReactiveTransactional
+    public Mono<LineagePojo> replaceLineagePaths(final List<LineagePojo> pojos) {
```
[the remark this answers is not in the document]
Comment (esmeneev) [reply odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveLineageRepositoryImpl.java]: Fixed by moving replaceLineagePaths to service.

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveLineageRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        final Set<String> establishers = pojos.stream()
+            .map(LineagePojo::getEstablisherOddrn)
+            .collect(Collectors.toSet());
+
+        DSL.deleteFrom(LINEAGE)
+            .where(LINEAGE.ESTABLISHER_ODDRN.in(establishers))
+            .execute();
```
[the remark this answers is not in the document]
Comment (esmeneev) [reply odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveLineageRepositoryImpl.java]: Fixed by moving replaceLineagePaths to service.

**Review by damirabdul — APPROVED** [review]

**Pull request merged**